refactor snapshotter

mhofman opened the pull request:
refs: https://github.com/Agoric/agoric-sdk/issues/6527
refs: https://github.com/Agoric/agoric-sdk/issues/8025 
refs: https://github.com/Agoric/agoric-sdk/issues/8031
Best reviewed commit-by-commit

## Description

In order to support the migrations needed for #8025 #8031, and to implement #6527, we need to enable using the snapshot pathways between golang and JS outside of the cosmos state-sync mechanism.

This PR splits the snapshotter into a `SwingStoreExportsHandler` module responsible for the communication and synchronization with JS, and a simplified cosmos `ExtensionSnapshotter` that uses the new SwingStoreExportsHandler hide the JS interactions.

Given the goroutine synchronization and multistep process of performing a SwingStore snapshot, there is a decent amount of interleaving needed, which is now expressed as facets provided as arguments to methods representing each step:

- The `SwingStoreExportsHandler`'s `InitiateExport()` takes a `SwingStoreExportEventHandler`, which has a `ExportInitiated()` method called from the goroutine it starts.
- `ExportInitiated()` is provided a `retrieveExport` callback. The event handler implemented by the `ExtensionSnapshotter` stores that callback and invokes it in its `SnapshotExtension()` method (internally invoked by the cosmos snapshot manager)
- `retrieveExport()` processes the data received from JS, and creates a `SwingStoreExportProvider` for it, which is provided as an argument to the `ExportRetrieved()` method of the `SwingStoreExportEventHandler`
- `ExportRetrieved()` is also implemented by the `ExtensionsSnapshotter` and uses that `provider` to recover the snapshot artifacts and write them out as state-sync payloads.

The new layering allows the `SwingStoreExportsHandler` to more easily notice errors happening during the retrieval phase, which are currently swallowed by cosmos's snapshot manager. It also introduces a way to explicitly await for the completion of a snapshot in progress, which will be needed for genesis export and the SwingStore shadow copy migration.

Finally the swing-store import/export options are threaded through all the way into the `SwingStoreExportsHandler` interface, allowing different kind of snapshot artifacts to be handled, as needed for the future use cases above.

### Security Considerations

None

### Scaling Considerations

None

### Documentation Considerations

None

### Testing Considerations

This is a refactor splitting an existing module. The tests covering that module were similarly split, and sometimes duplicated where appropriate.
With the split it may be possible to cover some more granular steps which could only be covered in aggregate before.

### Upgrade Considerations

This PR does not change any behavior or stored data, as such it does not have any upgrade impact.

mhofman commented on golang/cosmos/proto/agoric/swingset/swingset.proto:
```diff
@@ -150,8 +150,8 @@ message Egress {
 
-// The payload messages used by swingset state-sync 
-message ExtensionSnapshotterArtifactPayload {
+// The artifacts of swing-store snapshots
+message StoreSnapshotArtifact {
```
This type was only used internally, so it felt ok renaming. the protobuf breakage test is complaining, but it seems it's not a required test?

michaelfig replied:
You can silence the protobuf breakage CI with the `proto:expect-breakage` label.

mhofman replied:
Looks like the `expect-breakage` only influences mergify and not the test run itself.

mhofman commented:
I followed up with a lot of code comment clarifications / godoc fixes, so it may be easier to review if I'm allowed to rebase all the fixup commits. Any objections from reviewers?

gibson042 commented on golang/cosmos/x/swingset/keeper/extension_snapshotter.go:
```diff
@@ -0,0 +1,270 @@
+const SnapshotFormat = 1
+
+type activeExtensionSnapshot struct {
+	// The block height of the snapshot in progress
+	height uint64
+	// The logger for this snapshot
+	logger log.Logger
+	// The function to retrieve the exported artifacts
+	retrieveSwingStoreExport func() error
+	// The writer for extensions payloads
+	payloadWriter snapshots.ExtensionPayloadWriter
```
```suggestion
	// height is the block height of this in-progress snapshot.
	height uint64
	// logger is the destination for this snapshot's log messages.
	logger log.Logger
	// retrieveSwingStoreExport fetches this snapshot's exported artifacts from ???.
	retrieveSwingStoreExport func() error
	// payloadWriter writes bytes into the under-construction protobuf message for a snapshot.
	// It may be called multiple times, and often is (e.g., once per artifact).
	payloadWriter snapshots.ExtensionPayloadWriter
```

gibson042 commented on golang/cosmos/x/swingset/keeper/extension_snapshotter.go:
```diff
@@ -0,0 +1,270 @@
+	payloadWriter snapshots.ExtensionPayloadWriter
+}
+
+// SwingsetSnapshotter manages Swingset snapshots, ensuring insensitivity to
```
Is this a good place to document and/or reference external documentation of how a Swingset snapshot is structured and/or the general process for constructing one?

gibson042 commented on golang/cosmos/x/swingset/keeper/extension_snapshotter.go:
```diff
@@ -0,0 +1,270 @@
+// to ensure the results are deterministic.
+type SwingsetSnapshotter struct {
+	isConfigured            func() bool
+	takeAppSnapshot         func(height int64)
+	newRestoreContext       func(height int64) sdk.Context
```
I wonder if we should establish a convention to specifically use "blockHeight" rather than "height".

gibson042 commented on golang/cosmos/x/swingset/keeper/extension_snapshotter.go:
```diff
@@ -0,0 +1,270 @@
+}
+
+// NewSwingsetSnapshotter creates a new SwingsetSnapshotter
+func NewSwingsetSnapshotter(app *baseapp.BaseApp, swingStoreSnapshotter *SwingStoreSnapshotter, getSwingStoreExportData func(ctx sdk.Context) []*vstoragetypes.DataEntry) SwingsetSnapshotter {
```
The normal Go convention is for a NewFoo function to return *Foo (i.e., a _pointer_ to a Foo rather than a Foo itself). I don't know how rigid the convention is, so we may decide that this is fine, but on the other hand it's rather low effort to conform with it.

Likewise for the other New… functions.

mhofman replied:
I'm still very confused in golang as to what needs to be a pointer, and what can stay a struct value, and our codebase doesn't help clarify this confusion.

gibson042 commented on golang/cosmos/x/swingset/keeper/extension_snapshotter.go:
```diff
@@ -0,0 +1,270 @@
+// NewSwingsetSnapshotter creates a new SwingsetSnapshotter
+func NewSwingsetSnapshotter(app *baseapp.BaseApp, swingStoreSnapshotter *SwingStoreSnapshotter, getSwingStoreExportData func(ctx sdk.Context) []*vstoragetypes.DataEntry) SwingsetSnapshotter {
+	return SwingsetSnapshotter{
+		isConfigured:    func() bool { return app.SnapshotManager() != nil },
```
This seems like an inversion of responsibility... is there a way to refactor, or to somehow assert a connection between the returned value and this SwingsetSnapshotter?

mhofman replied:
This `isConfigured` was introduced for testability in an earlier PR (this is a simple split refactor). It doesn't strictly need to be there.

gibson042 commented on golang/cosmos/x/swingset/keeper/extension_snapshotter.go:
```diff
@@ -0,0 +1,270 @@
+
+// SnapshotFormat returns the default format the extension snapshotter uses to encode the
+// payloads when taking a snapshot.
+// It's defined within the extension, different from the global format for the whole state-sync snapshot.
```
I don't think I'm clear on the various levels of versioning, or e.g. if the value returned from SnapshotFormat corresponds with a single artifact or with the collection of all artifacts. Is it possible to improve this a bit?

mhofman replied:
This is inherited by the cosmos-sdk. Versioning applies to all "extensions payloads of this module" I believe.

gibson042 commented on golang/cosmos/x/swingset/keeper/extension_snapshotter.go:
```diff
@@ -0,0 +1,270 @@
+// InitiateSnapshot initiates a snapshot for the given height.
+// If a snapshot is already in progress, or if no snapshot manager is
+// configured, this will fail.
+// The snapshot operation is performed in a goroutine managed by the shared
+// SwingStoreSnapshotter. Use `WaitUntilSnapshotStarted` to synchronize commit
```
In what sense is the SwingStoreSnapshotter shared? What distinct entities reference a single one?

gibson042 commented on golang/cosmos/x/swingset/keeper/extension_snapshotter.go:
```diff
@@ -0,0 +1,270 @@
+func (snapshotter *SwingsetSnapshotter) InitiateSnapshot(height int64) error {
+	if !snapshotter.isConfigured() {
+		return fmt.Errorf("snapshot manager not configured")
+	}
```
```suggestion
	} else if height < 0 {
		return fmt.Errorf("block height must not be negative")
	}
```

gibson042 commented on golang/cosmos/x/swingset/keeper/extension_snapshotter.go:
```diff
@@ -0,0 +1,270 @@
+	})
+}
+
+// TakeSnapshot performs the app snapshot, which will call `SnapshotExtension`
```
Can the difference between "initiating" vs. "performing"/"taking" a snapshot be better documented? I wonder if the latter is even the right verb as opposed to something like "finish" or "populate".

gibson042 commented on golang/cosmos/x/swingset/keeper/extension_snapshotter.go:
```diff
@@ -0,0 +1,270 @@
+	}
+	snapshotter.activeExtensionSnapshot = &active
+
+	snapshotter.takeAppSnapshot(int64(height))
```
What's the benefit of apparently middle-layer-only conversion from int64 to uint64?

mhofman replied:
cosmos-sdk is itself inconsistent in its usage of uint64 vs int64. I've decided to settle on uint64. Unfortunately we're at a layer that straddles both.

gibson042 commented on golang/cosmos/x/swingset/keeper/extension_snapshotter.go:
```diff
@@ -0,0 +1,270 @@
+	return activeExtensionSnapshot.retrieveSwingStoreExport()
+}
+
+// ProcessExportedSnapshot handles the snapshot retrieved by the SwingStoreSnappshotter
```
```suggestion
// ProcessExportedSnapshot handles the snapshot retrieved by the SwingStoreSnapshotter
```

gibson042 commented on golang/cosmos/x/swingset/keeper/extension_snapshotter.go:
```diff
@@ -0,0 +1,270 @@
+// ProcessExportedSnapshot handles the snapshot retrieved by the SwingStoreSnappshotter
+// and writes it out to the SnapshotExtension's payloadWriter.
+// This operation is invoked by the SwingStoreSnapshotter nested in the goroutine
+// started by `InitiateSnapshot`.
```
I would really appreciate an all-in-one-place explanation of how these functions relate to each other in practice.

gibson042 commented on golang/cosmos/x/swingset/keeper/extension_snapshotter.go:
```diff
@@ -0,0 +1,270 @@
+	}
+
+	if activeExtensionSnapshot.height != provider.BlockHeight {
+		return fmt.Errorf("swingset snapshot received for unexpected height %d (expected %d)", provider.BlockHeight, activeExtensionSnapshot.height)
```
Is it a swing*set* snapshot that was received with the unexpected height, or a swing*store* snapshot?
```suggestion
		return fmt.Errorf("swingstore snapshot received for unexpected height %d (expected %d)", provider.BlockHeight, activeExtensionSnapshot.height)
```

gibson042 commented on golang/cosmos/x/swingset/keeper/extension_snapshotter.go:
```diff
@@ -0,0 +1,270 @@
+		return err
+	}
+
+	// Write out any retrieved export data as an untrusted artifact for debugging
```
```suggestion
	// For debugging, write out any retrieved export data as a single untrusted artifact
	// whose contents are a sequence of [key, value] JSON arrays,
	// each terminated by a line feed.
```

gibson042 commented on golang/cosmos/x/swingset/keeper/store_snapshotter.go:
```diff
@@ -0,0 +1,601 @@
+	"github.com/tendermint/tendermint/libs/log"
+)
+
+// ExportManifestFilename is the manifest filename which must be synchronized with the JS export/import tooling
```
What is the nature of the synchronization? Should there be a specific cross-reference here?

gibson042 commented on golang/cosmos/x/swingset/keeper/store_snapshotter.go:
```diff
@@ -0,0 +1,601 @@
+const UntrustedExportDataArtifactName = "UNTRUSTED-EXPORT-DATA"
+
+const exportDataFilename = "export-data.jsonl"
+const untrustedExportDataFilename = "untrusted-export-data.jsonl"
```
A note in case it doesn't come up later: untrusted export data is emitted as a sequence of line-feed-*terminated* JSON arrays, but https://jsonlines.org/ documents that lines are _separated_ and the validator at https://jsonlines.org/validator/ rejects input in which the last character is a line feed. The difference is subtle, but probably worth explicitly mentioning in e.g. golang/cosmos/x/swingset/keeper/store_snapshotter.go `RestoreSnapshot`.

mhofman replied:
ah yeah, this is line terminated json, which is fine in all our parsers, both in JS and golang. I'll make a note.

gibson042 commented on golang/cosmos/x/swingset/keeper/store_snapshotter.go:
```diff
@@ -0,0 +1,601 @@
+}
+
+type activeSnapshot struct {
+	// Whether the operation in progress is a restore
+	isRestore bool
+	// The block height of the snapshot in progress
+	height uint64
+	// The logger for this snapshot
+	logger log.Logger
+	// Use to synchronize the commit boundary
+	startedResult chan error
+	// Internal flag indicating whether the snapshot was retrieved
+	// Only read or written by the snapshot worker goroutine.
+	retrieved bool
+	// Closed when this snapshot is complete
+	done chan error
```
```suggestion
	// isRestore indicates whether the operation in progress is a restore.
	isRestore bool
	// height is the block height of this in-progress snapshot.
	height uint64
	// logger is the destination for this snapshot's log messages.
	logger log.Logger
	// startedResult is used to synchronize the commit boundary by ….
	startedResult chan error
	// retrieved is an internal flag indicating whether the snapshot was retrieved.
	// It must only be read or written by the snapshot worker goroutine.
	retrieved bool
	// done is a channel that is closed when this snapshot is complete.
	done chan error
```

gibson042 commented on golang/cosmos/x/swingset/keeper/store_snapshotter.go:
```diff
@@ -0,0 +1,601 @@
+}
+
+type exportManifest struct {
+	BlockHeight uint64 `json:"blockHeight,omitempty"`
+	// The filename of the export data
+	Data string `json:"data,omitempty"`
+	// The list of artifact names and their corresponding filenames
+	Artifacts [][2]string `json:"artifacts"`
```
```suggestion
	// BlockHeight is the block height of the manifest.
	BlockHeight uint64 `json:"blockHeight,omitempty"`
	// Data is the filename of the export data.
	Data string `json:"data,omitempty"`
	// Artifacts is the list of [artifact name, file name] pairs.
	Artifacts [][2]string `json:"artifacts"`
```

gibson042 commented on golang/cosmos/x/swingset/keeper/store_snapshotter.go:
```diff
@@ -0,0 +1,601 @@
+// The abstraction is similar to the JS side swing-store exporter abstraction,
+// but without the ability to list artifacts or random access them.
+type SwingStoreExportProvider struct {
+	// The block height of the SwingStore export
+	BlockHeight uint64
+	// A function to get the "export data" of the SwingStore export, if any
+	// errors with io.EOF if no "export data" is available for this export
+	GetExportData func() ([]*vstoragetypes.DataEntry, error)
+	// A function to retrieve each artifact in the SwingStore export
+	// errors with io.EOF once reaching the end of the artifact list
+	ReadArtifact func() (types.SwingStoreArtifact, error)
```
```suggestion
	// BlockHeight is the block height of the SwingStore export.
	BlockHeight uint64
	// GetExportData is a function to return the "export data" of the SwingStore export, if any.
	// It errors with io.EOF if the export contains no "export data".
	GetExportData func() ([]*vstoragetypes.DataEntry, error)
	// ReadArtifact is a function to return the next unread artifact in the SwingStore export.
	// It errors with io.EOF upon reaching the end of the artifact list.
	ReadArtifact func() (types.SwingStoreArtifact, error)
```

gibson042 commented on golang/cosmos/x/swingset/keeper/store_snapshotter.go:
```diff
@@ -0,0 +1,601 @@
+// creating a snapshot. It defines the interface of the object implementing
+// these operations, and which is provided to InitiateSnapshot.
+type SwingStoreSnapshotTaker interface {
+	// Called after the SwingStore snapshot was initiated in a goroutine
+	// Must call the `retrieveSwingStoreExport` function before returning, which
+	// will in turn synchronously invoke `ProcessExportedSnapshot` once the
+	// snapshot is ready
+	TakeSnapshot(height uint64, retrieveSwingStoreExport func() error) error
+	// Called when the SwingStore snapshot has been retrieved
+	// The implementation must synchronously consume the provider, which becomes
+	// invalid after the method returns.
+	ProcessExportedSnapshot(provider SwingStoreExportProvider) error
```
Can these doc comments be reworded to explicitly distinguish for each function:
* its purpose/effects/etc., and
* when it is expected to be invoked (and by what), and
* what requirements it must satisfy.

?

gibson042 commented on golang/cosmos/x/swingset/keeper/store_snapshotter.go:
```diff
@@ -0,0 +1,601 @@
+// times, but this is safe because when handling the messages, the JS side
+// does not perform operations affecting consensus and ignores state changes
+// since committing the previous block.
+// Some other `blockingSend` calls however do change the JS swing-store and
+// must happen before the Swingset controller on the JS side was inited.
```
How are the `blockingSend` calls that must happen before JS Swingset controller init distinguished here?

mhofman replied:
The `blockingSend` function takes a `mustNotBeInited` param?

gibson042 commented on golang/cosmos/x/swingset/keeper/store_snapshotter.go:
```diff
@@ -0,0 +1,601 @@
+	return nil
+}
+
+// InitiateSnapshot initiates a SwingStore snapshot for the given
+// height. If a snapshot is already in progress, this will fail. The snapshot
+// processing is delegated to the provided `snapshotTaker`.
```
```suggestion
// InitiateSnapshot synchronously verifies that there is not already a snapshot
// in progress and records the start of a new one for the provided block height,
// then launches a goroutine to asynchronously coordinate with JS for initiating
// and retrieving a snapshot, sending it to the provided `snapshotTaker` for processing,
// and ultimately releasing it.
```

That the behavior includes retrieval and processing also suggests to me that "initiate" is not the best name, although a similar argument would probably apply to using the name "take" for a function that synchronously starts a process which finishes asynchronously. It might be best to just reify the separation such that e.g. call sites look like
```go
if err := snapshotter.InitiateSnapshot(height); err != nil {
	…
}
go snapshotter.FinishSnapshot(taker, exportOptions)
```

mhofman replied:
That shape is not possible. Let's chat

gibson042 commented on golang/cosmos/x/swingset/keeper/store_snapshotter.go:
```diff
@@ -0,0 +1,601 @@
+				return errors.New("snapshot no longer active")
+			}
+
+			retrieveErr = snapshotter.retrieveSnapshot(snapshotTaker.ProcessExportedSnapshot)
```
Why does this function directly perform the `blockingSend`s for "initiate" and "discard" but not for "retrieve"?

mhofman replied:
Because of the way state-sync is implemented in the cosmos-sdk. Will try to better details the interleaving.

gibson042 commented on golang/cosmos/x/swingset/keeper/store_snapshotter.go:
```diff
@@ -0,0 +1,685 @@
+	"github.com/tendermint/tendermint/libs/log"
+)
+
+// This module abstract the handling of swing-store snapshots, also known as
```
```suggestion
// This module abstracts the handling of swing-store snapshots, also known as
```

gibson042 commented on golang/cosmos/x/swingset/keeper/store_snapshotter.go:
```diff
@@ -0,0 +1,685 @@
+// swing-store imports/exports, and the necessary communication with the
+// JS side.
+//
+// It's layering derives from the following requirements:
```
```suggestion
Its layering derives from the following requirements:
```

gibson042 commented on golang/cosmos/x/swingset/keeper/store_snapshotter.go:
```diff
@@ -0,0 +1,685 @@
+	"github.com/tendermint/tendermint/libs/log"
+)
+
+// This module abstract the handling of swing-store snapshots, also known as
+// swing-store imports/exports, and the necessary communication with the
+// JS side.
```
Do we have anywhere that describes how swing-store exports fit into the overall structure of cosmos state-sync snapshots? My recollection from our call is that there is some hook by which a module registers its integration(s) for cosmos to call when creating/loading/etc. a snapshot, and that the responsibility for creation is to produce an arbitrary number of protobuf "extension" entries with a common format version and for loading is to validate the format version of each extension entry and then consume it as appropriate.

gibson042 commented on golang/cosmos/x/swingset/keeper/store_snapshotter.go:
```diff
@@ -0,0 +1,685 @@
+// JS side.
+//
+// It's layering derives from the following requirements:
+// - Multiple golang components may perform swing-store snapshot operations,
+//   however a single operation must be active at a time, as defined by some
+//   global state on the JS side. There are no legitimate case of concurrent
+//   snapshot operations.
```
```suggestion
// - Multiple golang components may perform swing-store snapshotting,
//   but the JS side does not support concurrent snapshot operations.
```

gibson042 commented on golang/cosmos/x/swingset/keeper/store_snapshotter.go:
```diff
@@ -0,0 +1,685 @@
+//   however a single operation must be active at a time, as defined by some
+//   global state on the JS side. There are no legitimate case of concurrent
+//   snapshot operations.
+// - Some snapshot operations cannot block the main execution. In particular,
+//   cosmos's state-sync cannot block while taking a swing-store snapshot.
```
```suggestion
// - Long-running snapshot operations must not block the main execution.
//   In particular, cosmos state-sync cannot block while taking a swing-store
//   snapshot.
```

gibson042 commented on golang/cosmos/x/swingset/keeper/store_snapshotter.go:
```diff
@@ -0,0 +1,685 @@
+// has completed the restore before returning.
+
+// ExportManifestFilename is the manifest filename which must be synchronized with the JS export/import tooling
+// See packages/cosmic-swingset/src/export-kernel-db.js and packages/cosmic-swingset/src/import-kernel-db.js
+const ExportManifestFilename = "export-manifest.json"
```
There should be a similar cross-reference pointing here from packages/cosmic-swingset/src/export-kernel-db.js.

gibson042 commented on golang/cosmos/x/swingset/keeper/store_snapshotter.go:
```diff
@@ -0,0 +1,685 @@
+// not end up in the list of artifacts imported by the JS import tooling (which would fail).
+const UntrustedExportDataArtifactName = "UNTRUSTED-EXPORT-DATA"
+
+// For snapshot operations, the swing-store "export data" is exchanged with the
+// JS side as a file which encodes "export data" entries as a sequence of
+// [key, value] JSON arrays each terminated by a new line.
+// NB: this is not technically jsonlines since the entries are new line
+// terminated instead of being new line separated, however the parsers in both
+// JS and golang handle such extra whitespace.
```
:+1:

gibson042 commented on golang/cosmos/x/swingset/keeper/store_snapshotter.go:
```diff
@@ -0,0 +1,685 @@
+// creating a snapshot. It defines the interface of the object implementing
+// these operations, and which is provided to InitiateSnapshot.
+type SwingStoreSnapshotTaker interface {
+	// TakeSnapshot is called by InitiateSnapshot in a goroutine after the
+	// swing-store snapshot was initiated.
+	// This is where the component performing the snapshot must generate its result
+	// by performing its own off main thread snapshot operations, which include
+	// retrieving and processing the swing-store snapshot data.
+	//
+	// Must call the `retrieveSwingStoreExport` function before returning, which
+	// will in turn synchronously invoke `ProcessExportedSnapshot` once the
+	// swing-store snapshot is ready
```
:+1:

gibson042 commented on golang/cosmos/x/swingset/keeper/extension_snapshotter.go:
```diff
@@ -0,0 +1,288 @@
+
+var _ snapshots.ExtensionSnapshotter = &SwingsetSnapshotter{}
+
+// SnapshotFormat 1 is a proto message containing a swing-store artifact name, and its binary data
```
```suggestion
// SnapshotFormat 1 is a protobuf field value containing a swing-store artifact name and its binary data.
```

gibson042 commented on golang/cosmos/x/swingset/keeper/extension_snapshotter.go:
```diff
@@ -0,0 +1,288 @@
+	payloadWriter snapshots.ExtensionPayloadWriter
+}
+
+// SwingsetSnapshotter is the cosmos state-sync extension snapshotter for the x/swingset module
+// It handles the Swingset state that is not part of the Cosmos DB. Currently
+// that state is solely composed of the swing-store artifacts, as a copy of the
+// swing-store "export data" is streamed into the cosmos DB during execution.
+// When performing a snapshot, the extension leverages the SwingStoreSnapshotter
+// to retrieve the needed swing-store artifacts. When restoring a snapshot,
+// the extension combines the artifacts from the state-sync snapshot with the
+// swing-store "export data" from the already restored cosmos DB, to produce a
+// full swing-store snapshot that can be imported to create a new JS swing-store.
```
👍

gibson042 reviewed (approved):
As discussed in real time, I'd like to see a refactor in which the swing-store snapshotter communicates completion of its synchronous initiation to the main thread rather than directly driving the cosmos snapshot itself, which seems like an inversion of responsibility. But to reiterate, this need not be considered a blocking concern.

JimLarson reviewed (changes requested):
Giving up on the commit-by-commit review halfway through the 5th commit - too much back-and-forth - and  publishing the comments so far. I'm going to update to see recent commits and review the PR in its entirety.

mhofman commented on golang/cosmos/x/swingset/keeper/snapshotter.go:
```diff
@@ -387,12 +390,14 @@ func (snapshotter *SwingsetSnapshotter) RestoreExtension(height uint64, format u
 		return err
 	}
 
+	snapshotHeight := height
```
This is purely done to minimize the upcoming diff

mhofman commented on golang/cosmos/x/swingset/keeper/snapshotter.go:
```diff
@@ -446,6 +445,12 @@ func (snapshotter *SwingsetSnapshotter) RestoreExtension(height uint64, format u
+		if err != nil {
+			return err
+		}
+		exportDataFile.Close()
```
we're flushing the file, which I believe requires closing first?

mhofman commented on golang/cosmos/x/swingset/keeper/snapshotter.go:
```diff
@@ -465,8 +470,7 @@ func (snapshotter *SwingsetSnapshotter) RestoreExtension(height uint64, format u
 
-		switch {
-		case artifact.Name != UntrustedExportDataArtifactName:
+		if artifact.Name != UntrustedExportDataArtifactName {
```
I prefer keeping the main/normal case first

mhofman commented:
@michaelfig @JimLarson I believe I've addressed all feedback. It ended up being extensive changes, and introduces a couple conflicts with master, so right now I squashed everything and will rebase / relayer everything.

JimLarson commented on golang/cosmos/x/swingset/keeper/extension_snapshotter.go:
```diff
@@ -0,0 +1,297 @@
+	})
+}
+
+// ExportInitiated performs the actual cosmos state-sync app snapshot.
+// The cosmos implementation will ultimately call SnapshotExtension, which can
+// retrieve and process the SwingStore artifacts.
+// This method is invoked by the SwingStoreExportsHandler in a goroutine
+// started by InitiateExport, only if no other SwingStore export operation is
+// already in progress.
+//
+// Implements SwingStoreExportEventHandler
```
Comment before the rebase:

```suggestion
// TakeSnapshot sends a bridge message to request that Swingset perform a SwingStore export and
// then uses retrieveSwingStoreExport to retrieve it.
// This function will block until the export has been retrieved.
// No other SwingStore export may be in progress.
// Implements SwingStoreSnapshotTaker.
```

mhofman replied:
This suggestion is not accurate. I think this is rooted in the same confusion as @gibson042 regarding the inversion of responsibilities that the current layering suffers from. The rebase clarifies this, but in general I will follow up to clean up the layering and better assign responsibilities between the components, at which point I will also create a diagram to help visualize the flow.

JimLarson commented:
The rebase landed while I was in mid-review. I've submitted what was in-flight. Please address those comments as much as possible then please confirm that I have a stable target for review.

mhofman commented:
I have added a sequence diagram of the snapshot creation process: https://github.com/Agoric/agoric-sdk/blob/mhofman/refactor-snapshotter/docs/architecture/state-sync.md

gibson042 commented on golang/cosmos/proto/agoric/swingset/swingset.proto:
```diff
@@ -150,8 +150,8 @@ message Egress {
 
-// The payload messages used by swingset state-sync 
-message ExtensionSnapshotterArtifactPayload {
+// SwingStoreArtifact encodes the artifacts of a swing-store export.
```
```suggestion
// SwingStoreArtifact encodes an artifact of a swing-store export.
```

gibson042 commented on golang/cosmos/x/swingset/keeper/extension_snapshotter.go:
```diff
@@ -0,0 +1,305 @@
+	// retrieve the SwingStore's export provider which allows to read the export's
+	// artifacts used to populate this state-sync extension's payloads.
+	retrieveExport func() error
+	// payloadWriter is the callback provided by the state-sync snapshot manager
+	// to write the extension's payloads into the under-construction snapshot
+	// stream. It may be called multiple times, and often is (currently once per
+	// SwingStore export artifact).
+	payloadWriter snapshots.ExtensionPayloadWriter
```
```suggestion
	// payloadWriter is the callback provided by the state-sync snapshot manager
	// for an extension to write a payload into the under-construction snapshot
	// stream. It may be called multiple times, and often is (currently once per
	// SwingStore export artifact).
	payloadWriter snapshots.ExtensionPayloadWriter
```

gibson042 commented on golang/cosmos/x/swingset/keeper/swing_store_exports_handler.go:
```diff
@@ -0,0 +1,694 @@
+	nextArtifact := 0
+
+	readArtifact := func() (artifact types.SwingStoreArtifact, err error) {
+		if nextArtifact >= len(manifest.Artifacts) {
+			return artifact, io.EOF
+		}
```
```suggestion
		if nextArtifact == len(manifest.Artifacts) {
			return artifact, io.EOF
		} else if nextArtifact > len(manifest.Artifacts) {
			return artifact, fmt.Errorf("exceeded expected artifact count: %d > %d", nextArtifact, len(manifest.Artifacts))
		}
```

gibson042 reviewed (approved):
Skimmed changes since last review.

JimLarson commented on golang/cosmos/x/swingset/keeper/swing_store_exports_handler.go:
```diff
@@ -0,0 +1,696 @@
+	return nil
+}
+
+// retrieveExport retrieves an initiated export and calls exportRetrieved with
+// the retrieved export.
```
```suggestion
// retrieveExport retrieves an initiated export by sending a bridge message and on success
// calls exportRetrieved with the retrieved export.
//
// The bridge message returns a string giving the name of a directory containing the export
// files in a format described in <please insert reference here>.
//
// After calling exportRetrieved, the directory and its contents are deleted.
```

...and see below about renaming `exportRetrieved`.

JimLarson commented on golang/cosmos/x/swingset/keeper/swing_store_exports_handler.go:
```diff
@@ -0,0 +1,696 @@
+
+	blockHeight := provider.BlockHeight
+
+	// We technically don't need to create an active operation here since both
+	// InitiateExport and RestoreExport should only be called from the main
+	// thread, but it doesn't cost much to add in case things go wrong.
```
Cool!

JimLarson commented on golang/cosmos/x/swingset/keeper/swing_store_exports_handler.go:
```diff
@@ -0,0 +1,696 @@
+
+		if artifact.Name != UntrustedExportDataArtifactName {
+			// Artifact verifiable on import from the export data
+			// Since we cannot trust the state-sync artifact at this point, we generate
```
"cannot" or "can"? It's not the untrusted name, so shouldn't it be trusted?

mhofman commented on golang/cosmos/x/swingset/keeper/swing_store_exports_handler.go:
```diff
@@ -0,0 +1,696 @@
+	// BlockHeight is the block height of the SwingStore export.
+	BlockHeight uint64
+	// GetExportData is a function to return the "export data" of the SwingStore export, if any.
+	// It errors with io.EOF if the export contains no "export data".
```
Added a new commit to address this

JimLarson commented on golang/cosmos/x/swingset/keeper/extension_snapshotter.go:
```diff
@@ -0,0 +1,305 @@
+// This operation is invoked by the SwingStoreExportsHandler nested in the
+// goroutine started by InitiateExport.
+//
+// Implements SwingStoreExportEventHandler
```
The content of the swing store export is somewhat described, but not the format used for interchange between SwingSet and Cosmos. It's good enough for now, so I'll resolve this comment and we can improve later.

JimLarson commented on golang/cosmos/x/swingset/keeper/swing_store_exports_handler.go:
```diff
@@ -0,0 +1,794 @@
+	case <-operationDetails.exportDone:
+		activeOperation = nil
+	default:
+		// don't wait for it to finish
```
```suggestion
		// don't wait for it to finish
		// Note that in the case where there is no start error, this results in
		// a race in writing to or closing the exportDone channel and this
		// observation of it (see goroutine code in SwingStoreExportsHandler.InitiateExport()).
		// Thus this might result in a stale activeOperation sitting around to be seen
		// and cleared by the next call to this function.
		// This is harmless can cannot result in this function failing to wait for the
		// desired export to start, since SwingStoreExportHandler.InitiateExport()
		// will synchronously clear the stale activeOperation before returning.
```

JimLarson commented on golang/cosmos/x/swingset/keeper/swing_store_exports_handler.go:
```diff
@@ -0,0 +1,794 @@
+			return retrieveErr
+		})
+
+		// Restore any retrieve error swallowed by ExportStarted
```
```suggestion
		// Restore any retrieve error swallowed by OnExportStarted
```

JimLarson reviewed (approved):
LGTM at last! Thanks for applying all the previous feedback - I've verified that previous issues have been addressed to my satisfaction. Approval is conditional on applying the two comment suggestions here. One is trivial, the other explains why the dual-channel race (which is still present) is harmless.
